pyat/at/physics/linear.py

```diff
@@ -260,59 +288,144 @@ def _linopt(ring: Lattice, analyze, refpts=None, dp=None, dct=None, df=None,
+                dd0 = twin['ddispersion']
+                dalpha = twin['dalpha']
+                dbeta = twin['dbeta']
+            except (ValueError, KeyError):  # record arrays throw ValueError !
+                msg = ("'get_w' option for a line requires 'twiss_in' calculated "
+                       "with 'get_w' activated")
+                raise AtError(msg)
+
+            orbit = orbit + numpy.hstack((dd0, 1.0, 0.0)) * dp * dp
+            dorbit = numpy.hstack((d0+dd0*dp, 1.0, 0.0))
+
+            if (rmat is not None) and (drmat is not None):
+                rmat = rmat + drmat * dp
+            else:
+                alphas = alphas + dalpha * dp
+                betas = betas + dbeta * dp
+
+        if rmat is not None:
             # For some reason, "emittances" must be different...
-            sigm = twin['R'][0, ...]+10.0*twin['R'][1, ...]
-            if twin['R'].shape[0] >= 3:
-                sigm = sigm+0.1*twin['R'][2, ...]
-        except (ValueError, KeyError):  # record arrays throw ValueError !
+            sigm = rmat[0, ...]+10.0*rmat[1, ...]
+            if rmat.shape[0] >= 3:
+                sigm = sigm+0.1*rmat[2, ...]
+                dorbit = rmat[2, :, 4] / rmat[2, 4, 4, numpy.newaxis]
+        else:
             slices = [slice(2 * i, 2 * (i + 1)) for i in range(2)]
-            ab = numpy.stack((twin['alpha'], twin['beta']), axis=1)
+            ab = numpy.stack((alphas, betas), axis=1)
             sigm = numpy.zeros((4, 4))
             for slc, (alpha, beta) in zip(slices, ab):
                 gamma = (1.0+alpha*alpha)/beta
                 sigm[slc, slc] = numpy.array([[beta, -alpha], [-alpha, gamma]])
-        try:
-            d0 = twin['dispersion']
-        except (ValueError, KeyError):  # record arrays throw ValueError !
-            d0 = numpy.zeros((4,))
-        return orbit, sigm, d0
 
-    def chrom_w(ringup, ringdn, orbitup, orbitdn, refpts=None, **kwargs):
+        return orbit, sigm, dorbit
+
+    def chrom_w(ringup, ringdn, orbitup, orbitdn, twin=None,
+                refpts=None, **kwargs):
         """Compute the chromaticity and W-functions"""
         # noinspection PyShadowingNames
-        def off_momentum(rng, orb0):
-            mt, ms = get_matrix(rng, refpts=refpts, orbit=orb0, **kwargs)
-            vps, _, el0, els = analyze(mt, ms)
-            tunes = numpy.mod(numpy.angle(vps)/2.0/pi, 1.0)
-            return tunes, el0, els
-
-        def wget(ddp, elup, eldn):
+        def off_momentum(rng, orb0, dp=None, twin=None, **kwargs):
+            dp_step = kwargs.get('DPStep', DConstant.DPStep)
+            if twin is None:
+                mt, ms = get_matrix(rng, refpts=refpts, orbit=orb0, **kwargs)
+                mxx = mt
+                dpup = orb0[4] + 0.5 * dp_step
+                dpdn = orb0[4] - 0.5 * dp_step
+                o0up = None
+                o0dn = None
+            else:
+                orbit, sigma, dorbit = build_sigma(twiss_in, orb0, dp=dp)
```

Comment:
Here you are mixing a local variable `twin` and a non-local one `twiss_in` referring to the same thing. You should either use the local one everywhere within `off_momentum` (on this line), or simpler, since `twiss_in` is only used read-only, remove the `twin` local variables in `chrom_w` and `off_momentum` and access directly the non-local `twiss_in` here and on the test above.